internal/validation/gatewayclass_test.go

```diff
@@ -0,0 +1,165 @@
+			expect:    true,
+		},
+		{
+			name: "invalid name",
+			mutateNew: func(gc *gatewayapi_v1alpha1.GatewayClass) {
+				gc.Name = "invalid name"
+			},
+			expect: false,
+		},
+		{
+			name:      "existing admitted gatewayclass with same controller",
+			mutateNew: func(_ *gatewayapi_v1alpha1.GatewayClass) {},
+			mutateOld: func(gc *gatewayapi_v1alpha1.GatewayClass) {
+				gc.Spec.Controller = defaultController
+				gc.Status = gatewayapi_v1alpha1.GatewayClassStatus{
+					Conditions: []metav1.Condition{
+						{
+							Type:   string(gatewayapi_v1alpha1.GatewayClassConditionStatusAdmitted),
+							Status: metav1.ConditionTrue,
+						},
+					},
+				}
+			},
+			expect: false,
+		},
+		{
+			name:      "existing non-admitted gatewayclass with same controller",
+			mutateNew: func(_ *gatewayapi_v1alpha1.GatewayClass) {},
+			mutateOld: func(gc *gatewayapi_v1alpha1.GatewayClass) {
+				gc.Spec.Controller = defaultController
+				gc.Status = gatewayapi_v1alpha1.GatewayClassStatus{
+					Conditions: []metav1.Condition{
+						{
+							Type:   string(gatewayapi_v1alpha1.GatewayClassConditionStatusAdmitted),
+							Status: metav1.ConditionFalse,
+						},
+					},
+				}
+			},
+			expect: true,
+		},
+		{
+			name:      "existing gatewayclass with different controller",
+			mutateNew: func(_ *gatewayapi_v1alpha1.GatewayClass) {},
+			mutateOld: func(gc *gatewayapi_v1alpha1.GatewayClass) {
+				gc.Spec.Controller = "foo.io/bar"
+				gc.Status = gatewayapi_v1alpha1.GatewayClassStatus{
+					Conditions: []metav1.Condition{
+						{
+							Type:   string(gatewayapi_v1alpha1.GatewayClassConditionStatusAdmitted),
+							Status: metav1.ConditionTrue,
+						},
+					},
+				}
+			},
+			expect: true,
+		},
+		{
+			name: "invalid gatewayclass params",
+			mutateNew: func(gc *gatewayapi_v1alpha1.GatewayClass) {
+				gc.Name = "invalid-gatewayclass-params"
+				gc.Spec.ParametersRef = &gatewayapi_v1alpha1.ParametersReference{
+					Group: "foo",
+					Kind:  "bar",
+					Name:  "baz",
+				}
+			},
+			expect: false,
+		},
+	}
+
+	// Build the client
+	builder := fake.NewClientBuilder()
+	scheme, err := k8s.NewContourScheme()
+	if err != nil {
+		t.Fatalf("failed to build contour scheme: %v", err)
+	}
+	builder.WithScheme(scheme)
+
+	for _, tc := range testCases {
+		// Create the client
+		cl := builder.Build()
+
+		newMutated := new.DeepCopy()
+		tc.mutateNew(newMutated)
+
+		if err := cl.Create(context.TODO(), newMutated); err != nil {
+			t.Fatalf("failed to create gatewayclass: %v", err)
+		}
+
+		if tc.mutateOld != nil {
+			oldMutated := old.DeepCopy()
+			tc.mutateOld(oldMutated)
+
+			if err := cl.Create(context.TODO(), oldMutated); err != nil {
+				t.Fatalf("failed to create gatewayclass %q: %v", oldMutated.Name, err)
+			}
+		}
+
+		actual := ValidateGatewayClass(ctx, cl, newMutated, newMutated.Spec.Controller)
+		if actual != nil && tc.expect {
+			t.Fatalf("%q: expected %#v, got %v", tc.name, tc.expect, actual)
```

Comment:
could use `t.Run` to run these different cases and use the name of the tests instead of adding here